File: backend/app/reporting/exporters.py
"""Exportadores profesionales de la ruta crítica.

Genera, de forma determinista y sin dependencias externas:

* **MS Project (MSPDI XML)** — abrible en Microsoft Project.
* **Primavera P6 (XER)** — formato de intercambio de P6 (versión simplificada).
* **CSV** — compatible con Excel / hojas de cálculo.

Todos parten del cálculo CPM (fechas tempranas/tardías, holguras, camino crítico)
para que las tareas exportadas conserven su programación.
"""
from __future__ import annotations

import csv
import io
import xml.etree.ElementTree as ET
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple

from ..cpm import CPMEngine, Dependency, Task

HOURS_PER_DAY = 8

# Mapeo de tipos de dependencia a los códigos de cada herramienta.
_MSP_TYPE = {"FF": 0, "FS": 1, "SF": 2, "SS": 3}   # MS Project PredecessorLink.Type
_P6_TYPE = {"FS": "PR_FS", "SS": "PR_SS", "FF": "PR_FF", "SF": "PR_SF"}


@dataclass
class _Prepared:
    tasks: List[Task]
    result: object
    names: Dict[str, str]
    costs: Dict[str, float]
    progress: Dict[str, float]
    preds: Dict[str, List[Dependency]]     # successor_id -> [Dependency]
    milestones: Dict[str, bool]
    order: List[str]


def _prepare(payload: dict) -> Tuple[_Prepared, List[Dependency]]:
    raw_tasks = payload.get("tasks", [])
    if not raw_tasks:
        raise ValueError("No hay tareas para exportar.")
    tasks = [
        Task(
            id=str(t["id"]),
            duration=float(t.get("duration", 0) or 0),
            optimistic=_opt(t, "optimistic"),
            most_likely=_opt(t, "most_likely"),
            pessimistic=_opt(t, "pessimistic"),
        )
        for t in raw_tasks
    ]
    names = {str(t["id"]): str(t.get("name", t["id"])) for t in raw_tasks}
    costs = {str(t["id"]): float(t.get("planned_cost", t.get("cost", 0)) or 0) for t in raw_tasks}
    progress = {str(t["id"]): float(t.get("progress_pct", 0) or 0) for t in raw_tasks}
    milestones = {
        str(t["id"]): bool(t.get("is_milestone", False)) or float(t.get("duration", 0) or 0) == 0
        for t in raw_tasks
    }
    deps = [
        Dependency(
            predecessor=str(d["predecessor"]),
            successor=str(d["successor"]),
            dep_type=str(d.get("dep_type", "FS")),
            lag=float(d.get("lag", 0) or 0),
        )
        for d in payload.get("dependencies", [])
    ]
    result = CPMEngine(tasks, deps).compute()
    preds: Dict[str, List[Dependency]] = {t.id: [] for t in tasks}
    for d in deps:
        preds[d.successor].append(d)
    order = list(result.tasks.keys())
    return (
        _Prepared(tasks, result, names, costs, progress, preds, milestones, order),
        deps,
    )


def _opt(d: dict, key: str) -> Optional[float]:
    v = d.get(key)
    return float(v) if v is not None else None


# --------------------------------------------------------------------------- #
# MS Project (MSPDI XML)
# --------------------------------------------------------------------------- #
def to_msproject_xml(payload: dict) -> str:
    prep, _ = _prepare(payload)
    start_date = _parse_date(payload.get("start_date")) or datetime(2025, 1, 6)  # lunes
    project_name = str(payload.get("project_name", "Campo Crítico"))

    NS = "http://schemas.microsoft.com/project"
    ET.register_namespace("", NS)
    proj = ET.Element(f"{{{NS}}}Project")

    def sub(parent, tag, text):
        el = ET.SubElement(parent, f"{{{NS}}}{tag}")
        el.text = str(text)
        return el

    sub(proj, "Name", project_name)
    sub(proj, "StartDate", _iso(start_date))
    sub(proj, "DurationFormat", 7)         # 7 = días
    sub(proj, "CalendarUID", 1)

    tasks_el = ET.SubElement(proj, f"{{{NS}}}Tasks")
    uid = {tid: i + 1 for i, tid in enumerate(prep.order)}

    for i, tid in enumerate(prep.order):
        tr = prep.result.tasks[tid]
        t_el = ET.SubElement(tasks_el, f"{{{NS}}}Task")
        sub(t_el, "UID", uid[tid])
        sub(t_el, "ID", i + 1)
        sub(t_el, "Name", prep.names[tid])
        sub(t_el, "Active", 1)
        sub(t_el, "Manual", 0)
        sub(t_el, "Duration", _iso_duration(tr.duration))
        sub(t_el, "DurationFormat", 7)
        sub(t_el, "Start", _iso(start_date + timedelta(days=tr.early_start)))
        sub(t_el, "Finish", _iso(start_date + timedelta(days=tr.early_finish)))
        sub(t_el, "Milestone", 1 if prep.milestones[tid] else 0)
        sub(t_el, "PercentComplete", int(round(prep.progress[tid])))
        sub(t_el, "Critical", 1 if tr.is_critical else 0)
        sub(t_el, "TotalSlack", _iso_duration(tr.total_slack))
        for dep in prep.preds[tid]:
            link = ET.SubElement(t_el, f"{{{NS}}}PredecessorLink")
            sub(link, "PredecessorUID", uid[dep.predecessor])
            sub(link, "Type", _MSP_TYPE.get(dep.dep_type, 1))
            sub(link, "LinkLag", int(round(dep.lag * HOURS_PER_DAY * 60 * 10)))  # décimas de min
            sub(link, "LagFormat", 7)

    xml_bytes = ET.tostring(proj, encoding="utf-8", xml_declaration=True)
    return xml_bytes.decode("utf-8")


def _iso(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%dT%H:%M:%S")


def _iso_duration(days: float) -> str:
    """Duración MSPDI en ISO-8601 (horas), 8 h por día."""
    total_minutes = int(round(days * HOURS_PER_DAY * 60))
    h, m = divmod(total_minutes, 60)
    return f"PT{h}H{m}M0S"


def _parse_date(s) -> Optional[datetime]:
    if not s:
        return None
    try:
        return datetime.strptime(str(s)[:10], "%Y-%m-%d")
    except ValueError:
        return None


# --------------------------------------------------------------------------- #
# Primavera P6 (XER, simplificado)
# --------------------------------------------------------------------------- #
def to_p6_xer(payload: dict) -> str:
    prep, deps = _prepare(payload)
    start_date = _parse_date(payload.get("start_date")) or datetime(2025, 1, 6)
    project_id = str(payload.get("project_id", "CC1"))
    project_name = str(payload.get("project_name", "Campo Crítico"))
    now = datetime(2025, 1, 6, 0, 0, 0)

    lines: List[str] = []
    lines.append(
        "\t".join(
            ["ERMHDR", "18.8", _xer_date(now), "Project", "CampoDigital",
             "CampoDigital", "USD", "Campo Crítico Export"]
        )
    )

    def table(name: str, fields: List[str], rows: List[List[str]]):
        lines.append(f"%T\t{name}")
        lines.append("%F\t" + "\t".join(fields))
        for row in rows:
            lines.append("%R\t" + "\t".join(_xer_val(v) for v in row))

    # PROJECT
    table(
        "PROJECT",
        ["proj_id", "proj_short_name", "plan_start_date", "last_recalc_date"],
        [[project_id, project_name, _xer_date(start_date), _xer_date(now)]],
    )

    # TASK
    task_fields = [
        "task_id", "proj_id", "task_code", "task_name", "task_type",
        "target_drtn_hr_cnt", "phys_complete_pct", "early_start_date",
        "early_end_date", "total_float_hr_cnt", "driving_path_flag",
    ]
    task_rows = []
    for tid in prep.order:
        tr = prep.result.tasks[tid]
        task_rows.append([
            tid, project_id, tid, prep.names[tid],
            "TT_Mile" if prep.milestones[tid] else "TT_Task",
            round(tr.duration * HOURS_PER_DAY, 2),
            round(prep.progress[tid], 2),
            _xer_date(start_date + timedelta(days=tr.early_start)),
            _xer_date(start_date + timedelta(days=tr.early_finish)),
            round(tr.total_slack * HOURS_PER_DAY, 2),
            "Y" if tr.is_critical else "N",
        ])
    table("TASK", task_fields, task_rows)

    # TASKPRED (dependencias)
    pred_fields = ["task_pred_id", "task_id", "pred_task_id", "proj_id", "pred_type", "lag_hr_cnt"]
    pred_rows = []
    for i, d in enumerate(deps, start=1):
        pred_rows.append([
            i, d.successor, d.predecessor, project_id,
            _P6_TYPE.get(d.dep_type, "PR_FS"),
            round(d.lag * HOURS_PER_DAY, 2),
        ])
    table("TASKPRED", pred_fields, pred_rows)

    lines.append("%E")
    return "\n".join(lines) + "\n"


def _xer_date(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%d %H:%M")


def _xer_val(v) -> str:
    return "" if v is None else str(v)


# --------------------------------------------------------------------------- #
# CSV (Excel)
# --------------------------------------------------------------------------- #
def to_csv(payload: dict) -> str:
    prep, _ = _prepare(payload)
    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow([
        "ID", "Tarea", "Duración (d)", "Inicio temprano (ES)", "Fin temprano (EF)",
        "Inicio tardío (LS)", "Fin tardío (LF)", "Holgura total", "Crítica",
        "Predecesoras", "Avance %", "Costo planeado",
    ])
    for tid in prep.order:
        tr = prep.result.tasks[tid]
        preds = "; ".join(
            f"{d.predecessor}({d.dep_type}{('+' + str(_num(d.lag))) if d.lag else ''})"
            for d in prep.preds[tid]
        )
        writer.writerow([
            tid, prep.names[tid], _num(tr.duration), _num(tr.early_start),
            _num(tr.early_finish), _num(tr.late_start), _num(tr.late_finish),
            _num(tr.total_slack), "Sí" if tr.is_critical else "No",
            preds, _num(prep.progress[tid]), _num(prep.costs[tid]),
        ])
    return buf.getvalue()


def _num(x: float):
    return int(x) if float(x).is_integer() else round(x, 2)


# --------------------------------------------------------------------------- #
# Envoltorios para la API
# --------------------------------------------------------------------------- #
def export_msproject_from_dict(payload: dict) -> str:
    return to_msproject_xml(payload)


def export_p6_xer_from_dict(payload: dict) -> str:
    return to_p6_xer(payload)


def export_csv_from_dict(payload: dict) -> str:
    return to_csv(payload)

File: frontend/src/lib/exporters.js
/**
 * Exportadores — espejo en JavaScript de backend/app/reporting/exporters.py.
 * Genera CSV, MS Project (MSPDI XML) y Primavera P6 (XER) en el cliente
 * (funciona offline). Incluye utilidades de descarga y exportación a PNG.
 *
 * La generación de texto es determinista y equivalente al backend.
 */
import { computeCPM } from "./cpm.js";

const HOURS_PER_DAY = 8;
const MSP_TYPE = { FF: 0, FS: 1, SF: 2, SS: 3 };
const P6_TYPE = { FS: "PR_FS", SS: "PR_SS", FF: "PR_FF", SF: "PR_SF" };

const num = (x) => (Number.isInteger(x) ? x : Math.round(x * 100) / 100);

function prepare(payload) {
  const rawTasks = payload.tasks || [];
  if (!rawTasks.length) throw new Error("No hay tareas para exportar.");
  const engTasks = rawTasks.map((t) => ({
    id: String(t.id),
    duration: Number(t.duration || 0),
    optimistic: t.optimistic ?? null,
    most_likely: t.most_likely ?? null,
    pessimistic: t.pessimistic ?? null,
  }));
  const deps = (payload.dependencies || []).map((d) => ({
    predecessor: String(d.predecessor),
    successor: String(d.successor),
    dep_type: d.dep_type || "FS",
    lag: Number(d.lag ?? d.lag_days ?? 0),
  }));
  const result = computeCPM(engTasks, deps);
  const meta = {};
  rawTasks.forEach((t) => {
    meta[String(t.id)] = {
      name: t.name ?? String(t.id),
      cost: Number(t.planned_cost ?? t.cost ?? 0),
      progress: Number(t.progress_pct ?? 0),
      milestone: Boolean(t.is_milestone) || Number(t.duration || 0) === 0,
    };
  });
  const preds = {};
  Object.keys(result.tasks).forEach((id) => (preds[id] = []));
  deps.forEach((d) => preds[d.successor].push(d));
  return { result, meta, preds, order: Object.keys(result.tasks), deps };
}

/* --------------------------------- CSV ------------------------------------ */
export function toCsv(payload) {
  const { result, meta, preds, order } = prepare(payload);
  const esc = (v) => {
    const s = String(v ?? "");
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const rows = [[
    "ID", "Tarea", "Duración (d)", "Inicio temprano (ES)", "Fin temprano (EF)",
    "Inicio tardío (LS)", "Fin tardío (LF)", "Holgura total", "Crítica",
    "Predecesoras", "Avance %", "Costo planeado",
  ]];
  order.forEach((id) => {
    const tr = result.tasks[id];
    const m = meta[id];
    const p = preds[id]
      .map((d) => `${d.predecessor}(${d.dep_type}${d.lag ? "+" + num(d.lag) : ""})`)
      .join("; ");
    rows.push([
      id, m.name, num(tr.duration), num(tr.early_start), num(tr.early_finish),
      num(tr.late_start), num(tr.late_finish), num(tr.total_slack),
      tr.is_critical ? "Sí" : "No", p, num(m.progress), num(m.cost),
    ]);
  });
  return rows.map((r) => r.map(esc).join(",")).join("\r\n") + "\r\n";
}

/* ----------------------------- MS Project XML ----------------------------- */
function xmlEsc(s) {
  return String(s)
    .replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;").replace(/'/g, "&apos;");
}
function pad(n) {
  return String(n).padStart(2, "0");
}
function iso(date) {
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}T${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}`;
}
function addDays(base, days) {
  const d = new Date(base.getTime());
  d.setDate(d.getDate() + Math.round(days));
  return d;
}
function isoDuration(days) {
  const totalMin = Math.round(days * HOURS_PER_DAY * 60);
  const h = Math.floor(totalMin / 60);
  const m = totalMin % 60;
  return `PT${h}H${m}M0S`;
}

export function toMsProjectXml(payload) {
  const { result, meta, preds, order } = prepare(payload);
  const start = payload.start_date ? new Date(payload.start_date + "T00:00:00") : new Date(2025, 0, 6);
  const uid = {};
  order.forEach((id, i) => (uid[id] = i + 1));
  const parts = [];
  parts.push('<?xml version="1.0" encoding="utf-8"?>');
  parts.push('<Project xmlns="http://schemas.microsoft.com/project">');
  parts.push(`<Name>${xmlEsc(payload.project_name || "Campo Crítico")}</Name>`);
  parts.push(`<StartDate>${iso(start)}</StartDate>`);
  parts.push("<DurationFormat>7</DurationFormat><CalendarUID>1</CalendarUID>");
  parts.push("<Tasks>");
  order.forEach((id, i) => {
    const tr = result.tasks[id];
    const m = meta[id];
    parts.push("<Task>");
    parts.push(`<UID>${uid[id]}</UID><ID>${i + 1}</ID>`);
    parts.push(`<Name>${xmlEsc(m.name)}</Name>`);
    parts.push("<Active>1</Active><Manual>0</Manual>");
    parts.push(`<Duration>${isoDuration(tr.duration)}</Duration><DurationFormat>7</DurationFormat>`);
    parts.push(`<Start>${iso(addDays(start, tr.early_start))}</Start>`);
    parts.push(`<Finish>${iso(addDays(start, tr.early_finish))}</Finish>`);
    parts.push(`<Milestone>${m.milestone ? 1 : 0}</Milestone>`);
    parts.push(`<PercentComplete>${Math.round(m.progress)}</PercentComplete>`);
    parts.push(`<Critical>${tr.is_critical ? 1 : 0}</Critical>`);
    parts.push(`<TotalSlack>${isoDuration(tr.total_slack)}</TotalSlack>`);
    preds[id].forEach((d) => {
      parts.push("<PredecessorLink>");
      parts.push(`<PredecessorUID>${uid[d.predecessor]}</PredecessorUID>`);
      parts.push(`<Type>${MSP_TYPE[d.dep_type] ?? 1}</Type>`);
      parts.push(`<LinkLag>${Math.round(d.lag * HOURS_PER_DAY * 60 * 10)}</LinkLag><LagFormat>7</LagFormat>`);
      parts.push("</PredecessorLink>");
    });
    parts.push("</Task>");
  });
  parts.push("</Tasks></Project>");
  return parts.join("");
}

/* ------------------------------ Primavera XER ----------------------------- */
function xerDate(date) {
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())} ${pad(date.getHours())}:${pad(date.getMinutes())}`;
}

export function toP6Xer(payload) {
  const { result, meta, order, deps } = prepare(payload);
  const start = payload.start_date ? new Date(payload.start_date + "T00:00:00") : new Date(2025, 0, 6);
  const now = new Date(2025, 0, 6);
  const projId = payload.project_id || "CC1";
  const projName = payload.project_name || "Campo Crítico";
  const lines = [];
  lines.push(["ERMHDR", "18.8", xerDate(now), "Project", "CampoDigital", "CampoDigital", "USD", "Campo Crítico Export"].join("\t"));

  const table = (name, fields, rows) => {
    lines.push(`%T\t${name}`);
    lines.push("%F\t" + fields.join("\t"));
    rows.forEach((r) => lines.push("%R\t" + r.map((v) => (v == null ? "" : String(v))).join("\t")));
  };

  table("PROJECT", ["proj_id", "proj_short_name", "plan_start_date", "last_recalc_date"],
    [[projId, projName, xerDate(start), xerDate(now)]]);

  const taskRows = order.map((id) => {
    const tr = result.tasks[id];
    const m = meta[id];
    return [
      id, projId, id, m.name, m.milestone ? "TT_Mile" : "TT_Task",
      num(tr.duration * HOURS_PER_DAY), num(m.progress),
      xerDate(addDays(start, tr.early_start)), xerDate(addDays(start, tr.early_finish)),
      num(tr.total_slack * HOURS_PER_DAY), tr.is_critical ? "Y" : "N",
    ];
  });
  table("TASK",
    ["task_id", "proj_id", "task_code", "task_name", "task_type", "target_drtn_hr_cnt",
     "phys_complete_pct", "early_start_date", "early_end_date", "total_float_hr_cnt", "driving_path_flag"],
    taskRows);

  const predRows = deps.map((d, i) => [
    i + 1, d.successor, d.predecessor, projId, P6_TYPE[d.dep_type] || "PR_FS", num(d.lag * HOURS_PER_DAY),
  ]);
  table("TASKPRED", ["task_pred_id", "task_id", "pred_task_id", "proj_id", "pred_type", "lag_hr_cnt"], predRows);

  lines.push("%E");
  return lines.join("\n") + "\n";
}

/* --------------------------- Utilidades de navegador ---------------------- */
export function downloadText(filename, text, mime = "text/plain") {
  const blob = new Blob([text], { type: `${mime};charset=utf-8` });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

/** Exporta un elemento <svg> a PNG (data URL) y lo descarga. */
export function exportSvgToPng(svgEl, filename = "diagrama.png", scale = 2) {
  return new Promise((resolve, reject) => {
    const rect = svgEl.getBoundingClientRect();
    const w = rect.width || svgEl.viewBox.baseVal.width || 800;
    const h = rect.height || svgEl.viewBox.baseVal.height || 400;
    const clone = svgEl.cloneNode(true);
    clone.setAttribute("xmlns", "http://www.w3.org/2000/svg");
    const svgStr = new XMLSerializer().serializeToString(clone);
    const img = new Image();
    img.onload = () => {
      const canvas = document.createElement("canvas");
      canvas.width = w * scale;
      canvas.height = h * scale;
      const ctx = canvas.getContext("2d");
      ctx.fillStyle = "#ffffff";
      ctx.fillRect(0, 0, canvas.width, canvas.height);
      ctx.scale(scale, scale);
      ctx.drawImage(img, 0, 0, w, h);
      canvas.toBlob((blob) => {
        const url = URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.href = url;
        a.download = filename;
        a.click();
        URL.revokeObjectURL(url);
        resolve();
      }, "image/png");
    };
    img.onerror = reject;
    img.src = "data:image/svg+xml;base64," + btoa(unescape(encodeURIComponent(svgStr)));
  });
}
